### PR TITLE
contrib: Add lxc network setup for Bionic host

### PR DESCRIPTION
**Pull request by ken2812221**

### CODE REVIEW

File: contrib/gitian-build.py [hunk 1]
```diff
@@ -161,6 +167,9 @@ def main():
         os.environ['USE_DOCKER'] = '1'
     elif not args.kvm:
         os.environ['USE_LXC'] = '1'
+        if args.is_bionic:
+            os.environ['GITIAN_HOST_IP'] = '10.0.3.1'
```
Comment (MarcoFalke) [inline contrib/gitian-build.py]: Any reason why this conflicts the instructions in the docs repo?

https://github.com/bitcoin-core/docs/blame/fb356d71c1cb24da7560e19fa655f1022f7b7543/gitian-building/gitian-building-setup-gitian-debian.md#L39

Comment (ken2812221) [reply contrib/gitian-build.py]: The default setting in `/etc/default/lxc-net` is 10.0.3.1

**Comment by Sjors** [comment]
Concept ACK, this would help us make Gitian instructions shorter.

Nit: "automatical" -> "automatic", or: "automate setup for Ubuntu Bionic"

File: contrib/gitian-build.py (hunk 1 above)
Comment (ken2812221) [reply contrib/gitian-build.py]: Or should make it as the same as IP in document?

Comment (Sjors) [reply contrib/gitian-build.py]: I would stick to the default, unless there was a technical problem with it. The less customization the better. This part of the manual can be removed if this is merged. Maybe make sure that if `GITIAN_HOST_IP` is already set, you don't override it?

Comment (MarcoFalke) [reply contrib/gitian-build.py]: Agree with @Sjors

Comment (ken2812221) [reply contrib/gitian-build.py]: OK, I'll keep it default.

**Comment by ken2812221** [comment]
Update: Don't change `GITIAN_HOST_IP` and `LXC_GUEST_IP` if they exist in enviroment variable.

File: contrib/gitian-build.py [hunk 2]
```diff
@@ -11,10 +11,10 @@ def setup():
     if args.kvm:
         programs += ['python-vm-builder', 'qemu-kvm', 'qemu-utils']
     elif args.docker:
-        programs += ['docker.io']
+        programs += ['docker.io', 'docker-ce']
     else:
         programs += ['lxc', 'debootstrap']
-    subprocess.check_call(['sudo', 'apt-get', 'install', '-qq'] + programs)
+    subprocess.check_call(['sudo', 'apt-get', 'install', '-qq', '--ignore-missing'] + programs)
```
Comment (MarcoFalke) [inline contrib/gitian-build.py]: Looks like the `--ignore-missing` is not doing anything for me on bionic:

```
# apt install git docker-ce --ignore-missing
Reading package lists... Done
Building dependency tree       
Reading state information... Done
Package docker-ce is not available, but is referred to by another package.
This may mean that the package is missing, has been obsoleted, or
is only available from another source

E: Package 'docker-ce' has no installation candidate
# echo $?
100

File: contrib/gitian-build.py [hunk 3]
```diff
@@ -31,6 +31,10 @@ def setup():
         make_image_prog += ['--lxc']
     subprocess.check_call(make_image_prog)
     os.chdir(workdir)
+    if args.is_bionic:
+        subprocess.check_call(['sudo', 'sed', '-i', 's/lxcbr0/br0/', '/etc/default/lxc-net'])
+        print('Reboot is required')
+        exit(0)
```
Comment (MarcoFalke) [inline contrib/gitian-build.py]: Shouldn't this be only executed for lxc (and not docker/kvm)?

File: contrib/gitian-build.py (hunk 2 above)
Comment (ken2812221) [reply contrib/gitian-build.py]: Check if the OS is Bionic instead

File: contrib/gitian-build.py (hunk 3 above)
Comment (ken2812221) [reply contrib/gitian-build.py]: Yes.

File: contrib/gitian-build.py [hunk 4]
```diff
@@ -11,7 +11,7 @@ def setup():
     if args.kvm:
         programs += ['python-vm-builder', 'qemu-kvm', 'qemu-utils']
     elif args.docker:
-        programs += ['docker.io']
+        programs += ['docker.io' if args.is_bionic else 'docker-ce']
```
Comment (MarcoFalke) [inline contrib/gitian-build.py]: Sorry for nit-picking this so much, but the package is also named `docker.io` in xenial: https://packages.ubuntu.com/xenial/docker.io

And debian has neither docker.io, nor docker-ce, so I don't really see why this change is here?

File: contrib/gitian-build.py [hunk 5]
```diff
@@ -161,6 +167,11 @@ def main():
         os.environ['USE_DOCKER'] = '1'
     elif not args.kvm:
         os.environ['USE_LXC'] = '1'
+        if args.is_bionic:
+            if not 'GITIAN_HOST_IP' in os.environ.keys():
+                os.environ['GITIAN_HOST_IP'] = '10.0.3.1'
+            if not 'LXC_GUEST_IP' in os.environ.keys():
+                os.environ['LXC_GUEST_IP'] = '10.0.3.5'
```
Comment (MarcoFalke) [inline contrib/gitian-build.py]: These should also be exported on cosmic and later (and maybe debian?)

So no need to guard by `is_bionic`?

File: contrib/gitian-build.py (hunk 4 above)
Comment (ken2812221) [reply contrib/gitian-build.py]: Add an error message if no docker found

File: contrib/gitian-build.py (hunk 5 above)
Comment (ken2812221) [reply contrib/gitian-build.py]: Fixed

**Comment by MarcoFalke** [comment]
utACK 60c3c23a63bd7f681e0db86251ae0a372cb6b279

**Comment by fanquake** [comment]
utACK 60c3c23

File: contrib/gitian-build.py [hunk 6]
```diff
@@ -11,7 +11,14 @@ def setup():
     if args.kvm:
         programs += ['python-vm-builder', 'qemu-kvm', 'qemu-utils']
     elif args.docker:
-        programs += ['docker.io']
+        dockers = ['docker.io', 'docker-ce']
+        for i in dockers:
+            return_code = subprocess.call(['sudo', 'apt-get', 'install', '-qq', i])
```
Comment (fanquake) [inline contrib/gitian-build.py]: Had to lookup `-qq`
> Quiet. Produces output suitable for logging, omitting progress indicators. More q's will produce more quiet up to a maximum of two.

**Comment by MarcoFalke** [comment]
Could someone please update the recommended `*_IP` constants in https://github.com/bitcoin-core/docs as well?
